### ISSUE TITLE
blockheader::CidArrsEqual() can be wrong

### ISSUE THREAD

**Issue by arajasek** [filecoin-project/lotus#2857]
Probably not a big deal, but this method can return wrong values if there are duplicates:

```
func CidArrsEqual(a, b []cid.Cid) bool {
	if len(a) != len(b) {
		return false
	}

	// order ignoring compare...
	s := make(map[cid.Cid]bool)
	for _, c := range a {
		s[c] = true
	}

	for _, c := range b {
		if !s[c] {
			return false
		}
	}
	return true
}
```

This method would return true if called with `a = [X, Y, Z]` and `b = [X, X, X]`.

**Comment by schomatis** [comment]
Thanks for calling attention to this. This is _potentially_ a big deal. In my opinion this is the type of unnecessary risk we need to avoid.

The duplicate check (not sure how strictly enforced) is in `NewTipSet`. We normally use this comparison in the context of a `TipSetKey`. What we compare then is almost by definition a sorted array of CIDs, and in that case the logic should reflect this without other assumptions like the non-duplicates one here. If there is even a scenario where the CIDs are not sorted it should have its own function, decoupled from what's expected here.

**Comment by whyrusleeping** [comment]
To avoid sorting the inputs, or allocating new arrays, we should just do the following:
```go
for _, c := range a {
  var found bool
  for _, oc := range b {
    if c == oc {
      found = true
      break
    }
  }
  if !found {
    return false
  }
}
return true
```

roughly O(n^2 / 2), but input size is ~5 so who cares.

**Comment by schomatis** [comment]
But the inputs are already sorted, we should leverage that when applicable and make it explicit.

**Comment by whyrusleeping** [comment]
oh, no. because then [X, X, X] ,[X, Y, Z] would be incorrect. tricky.

**Comment by arajasek** [comment]
Yeah, we just assert A ⊆ B and B ⊆ A to conclude they're equal (the sizes are tiny)

**Comment by Reiers** [comment]
Hello everyone!

Looks like the fix is merged, and conversation died - close or keep open?

**Comment by schomatis** [comment]
Haven't been following up on this.

**Comment by Reiers** [comment]
Okay - I will leave it open for discussion. 
Thanks for the fast reply @schomatis !
